core/src/main/java/org/apache/calcite/rel/core/RelFactories.java

```diff
@@ -549,6 +555,34 @@ public RelNode createMatch(RelNode input, RexNode pattern,
           partitionKeys, orderKeys, interval);
     }
   }
+
+  /**
+   * Can create a {@link TableFunctionScan}
+   * of the appropriate type for a rule's calling convention.
+   */
+  public interface TableFunctionScanFactory {
+    /** Creates a {@link TableFunctionScan}. */
+    RelNode createTableFunctionScan(RelOptCluster cluster,
+        List<RelNode> inputs, RexNode rexCall, Type elementType,
```

Comment:
I don't think there should be an `inputs` parameter. You should pass 'int inputCount', and it will pop the right number of RelNodes off of the stack. See `union`, another relational operator that takes a variable number of inputs.

This impacts the `cursor` method. See my comments there.

Reply:
Note that it is a TableFunctionScanFactory. So I think it needs an `inputs` parameter.  But I think it may work in RelBuilder.